File: src/HarmoniHSE360.Domain/Entities/Hazard.cs
using HarmoniHSE360.Domain.Common;
using HarmoniHSE360.Domain.Events;
using HarmoniHSE360.Domain.ValueObjects;

namespace HarmoniHSE360.Domain.Entities;

public class Hazard : BaseEntity, IAuditableEntity
{
    public string Title { get; private set; } = string.Empty;
    public string Description { get; private set; } = string.Empty;
    public HazardCategory Category { get; private set; }
    public HazardType Type { get; private set; }
    public string Location { get; private set; } = string.Empty;
    public GeoLocation? GeoLocation { get; private set; }
    public HazardStatus Status { get; private set; }
    public HazardSeverity Severity { get; private set; }
    public DateTime IdentifiedDate { get; private set; }
    public DateTime? ExpectedResolutionDate { get; private set; }
    
    // Reporter information
    public int ReporterId { get; private set; }
    public User Reporter { get; private set; } = null!;
    public string ReporterDepartment { get; private set; } = string.Empty;
    
    // Risk assessment
    public int? CurrentRiskAssessmentId { get; private set; }
    public RiskAssessment? CurrentRiskAssessment { get; private set; }
    
    // Navigation properties
    private readonly List<HazardAttachment> _attachments = new();
    public IReadOnlyCollection<HazardAttachment> Attachments => _attachments.AsReadOnly();
    
    private readonly List<RiskAssessment> _riskAssessments = new();
    public IReadOnlyCollection<RiskAssessment> RiskAssessments => _riskAssessments.AsReadOnly();
    
    private readonly List<HazardMitigationAction> _mitigationActions = new();
    public IReadOnlyCollection<HazardMitigationAction> MitigationActions => _mitigationActions.AsReadOnly();
    
    private readonly List<HazardReassessment> _reassessments = new();
    public IReadOnlyCollection<HazardReassessment> Reassessments => _reassessments.AsReadOnly();

    // Audit fields
    public DateTime CreatedAt { get; private set; }
    public string CreatedBy { get; private set; } = string.Empty;
    public DateTime? LastModifiedAt { get; private set; }
    public string? LastModifiedBy { get; private set; }

    protected Hazard() { } // For EF Core

    public static Hazard Create(
        string title,
        string description,
        HazardCategory category,
        HazardType type,
        string location,
        HazardSeverity severity,
        int reporterId,
        string reporterDepartment,
        GeoLocation? geoLocation = null)
    {
        var hazard = new Hazard
        {
            Title = title,
            Description = description,
            Category = category,
            Type = type,
            Location = location,
            Severity = severity,
            Status = HazardStatus.Reported,
            IdentifiedDate = DateTime.UtcNow,
            ReporterId = reporterId,
            ReporterDepartment = reporterDepartment,
            GeoLocation = geoLocation,
            CreatedAt = DateTime.UtcNow,
            CreatedBy = $"User_{reporterId}"
        };

        hazard.AddDomainEvent(new HazardIdentifiedEvent(hazard));

        if (severity >= HazardSeverity.Major)
        {
            hazard.AddDomainEvent(new HighSeverityHazardIdentifiedEvent(hazard));
        }

        return hazard;
    }

    public void UpdateDetails(string title, string description, string modifiedBy)
    {
        Title = title;
        Description = description;
        LastModifiedAt = DateTime.UtcNow;
        LastModifiedBy = modifiedBy;

        AddDomainEvent(new HazardUpdatedEvent(this));
    }

    public void UpdateStatus(HazardStatus status, string modifiedBy)
    {
        var previousStatus = Status;
        Status = status;
        LastModifiedAt = DateTime.UtcNow;
        LastModifiedBy = modifiedBy;

        AddDomainEvent(new HazardStatusChangedEvent(this, previousStatus, status));

        if (status == HazardStatus.Resolved)
        {
            AddDomainEvent(new HazardResolvedEvent(this));
        }
    }

    public void UpdateSeverity(HazardSeverity severity, string modifiedBy)
    {
        var previousSeverity = Severity;
        Severity = severity;
        LastModifiedAt = DateTime.UtcNow;
        LastModifiedBy = modifiedBy;

        AddDomainEvent(new HazardSeverityChangedEvent(this, previousSeverity, severity));
    }

    public void SetGeoLocation(double latitude, double longitude)
    {
        GeoLocation = GeoLocation.Create(latitude, longitude);
    }

    public void SetExpectedResolutionDate(DateTime expectedResolutionDate)
    {
        ExpectedResolutionDate = expectedResolutionDate;
        LastModifiedAt = DateTime.UtcNow;
    }

    public void AddRiskAssessment(RiskAssessment riskAssessment)
    {
        _riskAssessments.Add(riskAssessment);
        
        // Set as current if it's the most recent active assessment
        if (riskAssessment.IsActive && 
            (CurrentRiskAssessment == null || riskAssessment.AssessmentDate > CurrentRiskAssessment.AssessmentDate))
        {
            CurrentRiskAssessmentId = riskAssessment.Id;
            CurrentRiskAssessment = riskAssessment;
        }

        AddDomainEvent(new RiskAssessmentCompletedEvent(this, riskAssessment));
    }

    public void AddMitigationAction(HazardMitigationAction action)
    {
        _mitigationActions.Add(action);
        
        // Update status if first mitigation action is added
        if (Status == HazardStatus.UnderAssessment)
        {
            Status = HazardStatus.Mitigating;
        }

        AddDomainEvent(new MitigationActionAddedEvent(this, action));
    }

    public void AddAttachment(string fileName, string filePath, long fileSize, string uploadedBy)
    {
        var attachment = new HazardAttachment(Id, fileName, filePath, fileSize, uploadedBy);
        _attachments.Add(attachment);
    }

    public void ScheduleReassessment(DateTime nextAssessmentDate, string reason)
    {
        var reassessment = new HazardReassessment(Id, nextAssessmentDate, reason);
        _reassessments.Add(reassessment);

        AddDomainEvent(new HazardReassessmentScheduledEvent(this, nextAssessmentDate, reason));
    }

    public void Close(string closureNotes, string closedBy)
    {
        // Check if all critical mitigation actions are completed
        var pendingCriticalActions = _mitigationActions
            .Where(ma => ma.Status != MitigationActionStatus.Completed && 
                        ma.Type == MitigationActionType.Elimination)
            .ToList();

        if (pendingCriticalActions.Any())
        {
            throw new InvalidOperationException("Cannot close hazard with pending critical mitigation actions");
        }

        Status = HazardStatus.Closed;
        LastModifiedAt = DateTime.UtcNow;
        LastModifiedBy = closedBy;

        AddDomainEvent(new HazardClosedEvent(this, closureNotes));
    }
}

public enum HazardCategory
{
    Physical = 1,
    Chemical = 2,
    Biological = 3,
    Ergonomic = 4,
    Psychological = 5,
    Environmental = 6,
    Fire = 7,
    Electrical = 8,
    Mechanical = 9,
    Radiation = 10
}

public enum HazardType
{
    Slip = 1,
    Trip = 2,
    Fall = 3,
    Cut = 4,
    Burn = 5,
    Exposure = 6,
    Collision = 7,
    Entrapment = 8,
    Explosion = 9,
    Fire = 10,
    Other = 99
}

public enum HazardStatus
{
    Reported = 1,
    UnderAssessment = 2,
    ActionRequired = 3,
    Mitigating = 4,
    Monitoring = 5,
    Resolved = 6,
    Closed = 7
}

public enum HazardSeverity
{
    Negligible = 1,
    Minor = 2,
    Moderate = 3,
    Major = 4,
    Catastrophic = 5
}

File: src/HarmoniHSE360.Web/ClientApp/src/types/hazard.ts
// User types (shared)
export interface UserDto {
  id: number;
  name: string;
  email: string;
  department?: string;
  position?: string;
  employeeId?: string;
}

// Core Hazard types
export interface HazardDto {
  id: number;
  title: string;
  description: string;
  category: string;
  type: string;
  location: string;
  latitude?: number;
  longitude?: number;
  status: string;
  severity: string;
  identifiedDate: string;
  expectedResolutionDate?: string;
  
  // Reporter info
  reporterName: string;
  reporterEmail?: string;
  reporterDepartment: string;
  
  // Current risk assessment summary
  currentRiskLevel?: string;
  currentRiskScore?: number;
  lastAssessmentDate?: string;
  
  // Related counts
  attachmentsCount: number;
  riskAssessmentsCount: number;
  mitigationActionsCount: number;
  pendingActionsCount: number;
  
  // Audit fields
  createdAt: string;
  createdBy?: string;
  lastModifiedAt?: string;
  lastModifiedBy?: string;
  
  // Optional detailed objects for expanded views
  reporter?: UserDto;
  currentRiskAssessment?: RiskAssessmentDto;
}

export interface HazardDetailDto extends HazardDto {
  attachments: HazardAttachmentDto[];
  riskAssessments: RiskAssessmentDto[];
  mitigationActions: HazardMitigationActionDto[];
  reassessments: HazardReassessmentDto[];
}

export interface HazardAttachmentDto {
  id: number;
  fileName: string;
  fileSize: number;
  contentType: string;
  uploadedBy: string;
  uploadedAt: string;
  description?: string;
  downloadUrl: string;
}

export interface RiskAssessmentDto {
  id: number;
  hazardId: number;
  type: string;
  assessorName: string;
  assessmentDate: string;
  
  // Risk scoring
  probabilityScore: number;
  severityScore: number;
  riskScore: number;
  riskLevel: string;
  
  // Assessment details
  potentialConsequences: string;
  existingControls: string;
  recommendedActions: string;
  additionalNotes: string;
  
  // Review cycle
  nextReviewDate: string;
  isActive: boolean;
  
  // Approval
  isApproved: boolean;
  approvedByName?: string;
  approvedAt?: string;
  approvalNotes?: string;
  
  // Optional detailed objects
  assessor?: UserDto;
  approvedBy?: UserDto;
}

export interface HazardMitigationActionDto {
  id: number;
  hazardId: number;
  actionDescription: string;
  type: string;
  status: string;
  priority: string;
  targetDate: string;
  completedDate?: string;
  assignedToName: string;
  completionNotes?: string;
  
  // Cost tracking
  estimatedCost?: number;
  actualCost?: number;
  
  // Effectiveness
  effectivenessRating?: number;
  effectivenessNotes?: string;
  
  // Verification
  requiresVerification: boolean;
  verifiedByName?: string;
  verifiedAt?: string;
  verificationNotes?: string;
  
  // Optional detailed objects
  assignedTo?: UserDto;
  verifiedBy?: UserDto;
}

export interface HazardReassessmentDto {
  id: number;
  hazardId: number;
  scheduledDate: string;
  reason: string;
  isCompleted: boolean;
  completedAt?: string;
  completedByName?: string;
  completionNotes?: string;
  createdAt: string;
  
  // Optional detailed objects
  completedBy?: UserDto;
}

// Request types
export interface CreateHazardRequest {
  title: string;
  description: string;
  category: string;
  type: string;
  location: string;
  latitude?: number;
  longitude?: number;
  severity: string;
  reporterId: number;
  reporterDepartment: string;
  expectedResolutionDate?: string;
  attachments?: File[];
}

export interface UpdateHazardRequest {
  id: number;
  title: string;
  description: string;
  category: string;
  type: string;
  location: string;
  latitude?: number;
  longitude?: number;
  status: string;
  severity: string;
  expectedResolutionDate?: string;
  statusChangeReason?: string;
  newAttachments?: File[];
  attachmentsToRemove?: number[];
}

export interface CreateRiskAssessmentRequest {
  hazardId: number;
  type: string;
  assessorId: number;
  probabilityScore: number;
  severityScore: number;
  potentialConsequences: string;
  existingControls: string;
  recommendedActions: string;
  additionalNotes?: string;
  nextReviewDate: string;
  setAsCurrent?: boolean;
}

export interface CreateMitigationActionRequest {
  hazardId: number;
  actionDescription: string;
  type: string;
  priority: string;
  targetDate: string;
  assignedToId: number;
  estimatedCost?: number;
  requiresVerification?: boolean;
  notes?: string;
}

// Query types
export interface GetHazardsParams {
  pageNumber?: number;
  pageSize?: number;
  searchTerm?: string;
  category?: string;
  type?: string;
  status?: string;
  severity?: string;
  riskLevel?: string;
  location?: string;
  department?: string;
  reporterId?: number;
  assignedToId?: number;
  identifiedDateFrom?: string;
  identifiedDateTo?: string;
  expectedResolutionDateFrom?: string;
  expectedResolutionDateTo?: string;
  latitude?: number;
  longitude?: number;
  radiusKm?: number;
  sortBy?: string;
  sortDirection?: string;
  includeAttachments?: boolean;
  includeRiskAssessments?: boolean;
  includeMitigationActions?: boolean;
  includeReporter?: boolean;
  onlyUnassessed?: boolean;
  onlyOverdue?: boolean;
  onlyHighRisk?: boolean;
  onlyMyHazards?: boolean;
}

export interface GetHazardsResponse {
  hazards: HazardDto[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  summary: HazardsSummary;
}

export interface HazardsSummary {
  totalHazards: number;
  openHazards: number;
  highRiskHazards: number;
  overdueActions: number;
  unassessedHazards: number;
  hazardsByCategory: Record<string, number>;
  hazardsBySeverity: Record<string, number>;
  hazardsByStatus: Record<string, number>;
}

// Dashboard types
export interface HazardDashboardDto {
  overview: HazardOverviewMetrics;
  riskAnalysis: RiskMetrics;
  performance: PerformanceMetrics;
  trends: TrendAnalytics;
  locationData: LocationAnalytics;
  compliance: ComplianceMetrics;
  recentActivities: RecentActivityDto[];
  alerts: HazardAlertDto[];
}

export interface HazardOverviewMetrics {
  totalHazards: number;
  openHazards: number;
  resolvedHazards: number;
  highRiskHazards: number;
  criticalRiskHazards: number;
  overdueActions: number;
  unassessedHazards: number;
  newHazardsThisMonth: number;
  totalHazardsChange: number;
  highRiskChange: number;
  resolutionRateChange: number;
}

export interface RiskMetrics {
  riskLevelDistribution: Record<string, number>;
  categoryDistribution: Record<string, number>;
  severityDistribution: Record<string, number>;
  averageRiskScore: number;
  riskAssessmentsCompleted: number;
  riskAssessmentsPending: number;
  riskAssessmentCompletionRate: number;
}

export interface PerformanceMetrics {
  averageResolutionTime: number;
  mitigationActionCompletionRate: number;
  totalMitigationActions: number;
  completedMitigationActions: number;
  overdueMitigationActions: number;
  averageActionEffectiveness: number;
  costSavingsFromMitigation: number;
  actionTypeDistribution: Record<string, number>;
}

export interface TrendAnalytics {
  hazardReportingTrend: DataPointDto[];
  riskLevelTrend: DataPointDto[];
  resolutionTimeTrend: DataPointDto[];
  categoryTrend: DataPointDto[];
  trendDirection: string;
  keyInsights: string[];
}

export interface LocationAnalytics {
  hotspotLocations: HazardLocationDto[];
  departmentDistribution: Record<string, number>;
  geographicClusters: GeographicClusterDto[];
  mostAffectedArea: string;
  locationsWithHazards: number;
}

export interface ComplianceMetrics {
  overallComplianceScore: number;
  complianceViolations: number;
  auditFindings: number;
  regulatoryReportingCompliance: number;
  nonComplianceAreas: string[];
  lastComplianceReview?: string;
  nextComplianceReview?: string;
}

export interface RecentActivityDto {
  id: number;
  activityType: string;
  title: string;
  description: string;
  timestamp: string;
  performedBy: string;
  severity: string;
  relatedEntityId: number;
}

export interface HazardAlertDto {
  id: number;
  alertType: string;
  title: string;
  message: string;
  severity: string;
  createdAt: string;
  hazardId: number;
  hazardTitle: string;
  isAcknowledged: boolean;
}

export interface DataPointDto {
  label: string;
  value: number;
  secondaryValue?: string;
}

export interface HazardLocationDto {
  location: string;
  hazardCount: number;
  highRiskCount: number;
  latitude?: number;
  longitude?: number;
  department: string;
}

export interface GeographicClusterDto {
  centerLatitude: number;
  centerLongitude: number;
  hazardCount: number;
  radiusMeters: number;
  primaryCategory: string;
  riskLevel: string;
}

export interface GetHazardDashboardParams {
  dateFrom?: string;
  dateTo?: string;
  department?: string;
  includeTrends?: boolean;
  includeLocationAnalytics?: boolean;
  includeComplianceMetrics?: boolean;
  includePerformanceMetrics?: boolean;
  personalizedView?: boolean;
}

// Constants for dropdowns
export const HAZARD_CATEGORIES = [
  'Physical',
  'Chemical', 
  'Biological',
  'Ergonomic',
  'Psychological',
  'Environmental',
  'Fire',
  'Electrical',
  'Mechanical',
  'Radiation'
] as const;

export const HAZARD_TYPES = [
  'Slip',
  'Trip', 
  'Fall',
  'Cut',
  'Burn',
  'Exposure',
  'Collision',
  'Entrapment',
  'Explosion',
  'Fire',
  'Other'
] as const;

export const HAZARD_STATUSES = [
  'Reported',
  'UnderAssessment',
  'ActionRequired',
  'Mitigating',
  'Monitoring',
  'Resolved',
  'Closed'
] as const;

export const HAZARD_SEVERITIES = [
  'Negligible',
  'Minor',
  'Moderate',
  'Major',
  'Catastrophic'
] as const;

export const RISK_LEVELS = [
  'VeryLow',
  'Low',
  'Medium',
  'High',
  'Critical'
] as const;

export const RISK_ASSESSMENT_TYPES = [
  'General',
  'JSA',
  'HIRA',
  'Environmental',
  'Fire'
] as const;

export const MITIGATION_ACTION_TYPES = [
  'Elimination',
  'Substitution',
  'Engineering',
  'Administrative',
  'PPE'
] as const;

export const MITIGATION_PRIORITIES = [
  'Low',
  'Medium',
  'High',
  'Critical'
] as const;